### PR TITLE
merging

### PR DESCRIPTION
**Pull request by matteofabbri**
fixed branch problem

### CODE REVIEW

File: src/Mages.Core/Ast/ExpressionParser.cs [hunk 1]
```diff
@@ -72,30 +73,17 @@ private IStatement ParseKeywordStatement(IEnumerator<IToken> tokens)
...
-            }
-            else if (current.Is(Keywords.Break))
-            {
-                return ParseBreakStatement(tokens);
-            }
-            else if (current.Is(Keywords.Continue))
-            {
-                return ParseContinueStatement(tokens);
-            }
+            if (current.Is(Keywords.Var)) return ParseVarStatement(tokens);
```
Comment (FlorianRappl) [inline src/Mages.Core/Ast/ExpressionParser.cs]: Does not follow code convention.

File: src/Mages.Core.Tests/PluginTests.cs [hunk 2]
```diff
@@ -62,7 +62,7 @@ public void RemovingInexistingPluginHasNoEffect()
             content["a"] = new Function(args => (Double)args.Length);
             var plugin = new Plugin(metaData, content);
             var engine = new Engine();
-            engine.Globals["a"] = content["a"];
```
Comment (FlorianRappl) [inline src/Mages.Core.Tests/PluginTests.cs]: Plugin should be available at API layer.

File: src/Mages.Core.Tests/ExtensibilityTests.cs [hunk 3]
```diff
@@ -149,7 +149,7 @@ public void ReplacingAnExistingFunctionOverwrites()
         public void AddingAGlobalConstant()
         {
             var engine = new Engine();
-            engine.Globals["Answer"] = 42.0;
+            engine.Scope["Answer"] = 42.0;
```
Comment (FlorianRappl) [inline src/Mages.Core.Tests/ExtensibilityTests.cs]: Constant should be available at API layer.

File: src/Mages.Core.Tests/IntellisenseTests.cs [hunk 4]
```diff
@@ -12,7 +12,7 @@ public void EmptyScopeAndSourceYieldsKeywords()
         {
             var source = "";
             var engine = new Engine();
-            engine.Globals.Clear();
+            engine.Scope.Clear();
```
Comment (FlorianRappl) [inline src/Mages.Core.Tests/IntellisenseTests.cs]: We want to clear the globals (API layer) - not the scope. The scope (user layer) is initially empty!

File: src/Mages.Core/EngineExtensions.cs [hunk 5]
```diff
@@ -149,7 +149,7 @@ public static IEnumerable<VariableExpression> FindMissingSymbols(this Engine eng
                 var symbol = symbols[i];
                 var name = symbol.Name;
 
-                if (engine.Scope.ContainsKey(name) || engine.Globals.ContainsKey(name))
+                if (engine.Scope.ContainsKey(name) || engine.Scope.ContainsKey(name))
```
Comment (FlorianRappl) [inline src/Mages.Core/EngineExtensions.cs]: Redundancy. (but invalid anyway)

File: src/Mages.Core/Engine.cs [hunk 6]
```diff
@@ -51,20 +51,12 @@ public IParser Parser
...
         /// Gets the used global scope.
         /// </summary>
-        public IDictionary<String, Object> Scope
+        public Scope Scope
         {
             get { return _scope; }
         }
 
         /// <summary>
-        /// Gets the used global function layer.
```
Comment (FlorianRappl) [inline src/Mages.Core/Engine.cs]: Breaking change -- not allowed.

File: src/Mages.Core/Mages.Core.csproj [hunk 7]
```diff
@@ -121,7 +121,6 @@
     <Compile Include="Runtime\EntryChangedArgs.cs" />
     <Compile Include="Runtime\Functions\AttachedProperties.cs" />
     <Compile Include="Runtime\Functions\Curry.cs" />
-    <Compile Include="Runtime\Functions\Global.cs" />
```
Comment (FlorianRappl) [inline src/Mages.Core/Mages.Core.csproj]: Why do you remove the standard functions?!

File: src/Mages.Core/Runtime/GlobalScope.cs [hunk 8]
```diff
@@ -3,17 +3,46 @@
...
-            : base(scope ?? new Dictionary<String, Object>(), new Dictionary<String, Object>(Global.Mapping))
+        void LoadType(Type type)
+        {
+            foreach (var f in type.GetFields(BindingFlags.Public | BindingFlags.Static))
+                _scope.Add(f.Name.ToLower(), f.GetValue(null));
+        }
+
+        public GlobalScope(IDictionary<String, Object> scope): base(null)
+        {
+            if (scope == null)
```
Comment (FlorianRappl) [inline src/Mages.Core/Runtime/GlobalScope.cs]: Nope, too expensive. Also I still want to be able to fine-tune this. Therefore the Global.cs was there intentionally. It is *way* cheaper to lead and much easier to control.

**Comment by FlorianRappl** [comment]
While I want to merge this in some sense, generally there are way too many issues here.

* This introduces breaking changes, i.e., would be v2 of MAGES
* It collides with *many* design decisions, e.g., taking globals explicitly, prefer to expose .NET types over custom types
* The difference between API and user layer is important (as is the distinction between a local and a global scope)
* The fact that existing unit tests have been modified is already an indicator of breaking changes / a conflict in the decisions

We can discuss some of these points, but at least an item (issue) beforehand is required. Otherwise this is not transparent. I know that this all may look picky, but be assured that some QA is required in order to get this as polished and consistent as possible.

File: src/Mages.Core/Runtime/Stringify.cs [hunk 9]
```diff
@@ -1,192 +1,185 @@
-﻿namespace Mages.Core.Runtime
```
Comment (FlorianRappl) [inline src/Mages.Core/Runtime/Stringify.cs]: Again this is a full changeset and not just the desired partial diff. Check: Encoding, file endings, and other global settings that essentially swap the full file.

File: src/Mages.Core/Runtime/Functions/StandardFunctions.cs [hunk 10]
```diff
@@ -577,54 +569,12 @@ public static class StandardFunctions
...
-            }
-
-            if (args.Length == 2 && args[0] is IEnumerable && args[1] is Function)
-            {
-                var enu = args[0] as IEnumerable;
-                var f = args[1] as Function;
-
-                return enu.YieldAround(item => f.Invoke(new[] { item }));
-            }
-            return null;
```
Comment (FlorianRappl) [inline src/Mages.Core/Runtime/Functions/StandardFunctions.cs]: What about currying? If args.Length == 0 the function should be curried like all others.

File: src/Mages.Core/Runtime/Functions/StandardFunctions.cs [hunk 11]
```diff
@@ -659,6 +659,7 @@ public static class StandardFunctions
         {
             return Curry.MinTwo(Where, args) ??
                 If.Is<Function, Double[,]>(args, (f, m) => m.Where(f)) ??
+                If.Is<Function, IEnumerable>(args, (f, enu) => enu.Where(f)) ??
```
Comment (FlorianRappl) [inline src/Mages.Core/Runtime/Functions/StandardFunctions.cs]: This one will always leave out the dictionary, as the dictionary implements an `IEnumerable`.

File: src/Mages.Core/Mages.Core.csproj (hunk 7 above)
Comment (matteofabbri) [inline src/Mages.Core/Mages.Core.csproj]: they aren't removed, just automatically loaded into globalscope via reflection

File: src/Mages.Core/Ast/ExpressionParser.cs (hunk 1 above)
Comment (matteofabbri) [inline src/Mages.Core/Ast/ExpressionParser.cs]: ok.

File: src/Mages.Core/Runtime/Functions/StandardFunctions.cs (hunk 11 above)
Comment (matteofabbri) [inline src/Mages.Core/Runtime/Functions/StandardFunctions.cs]: yep i fixed it the next commit

**Comment by matteofabbri** [comment]
"The difference between API and user layer is important (as is the distinction between a local and a global scope)"

When i looked at Engine.Globals and Engine.Scope i saw some redundancy, likes a global scope splitted in two parts, but probably i just didn't understood something, can you please explain me better what do you mean?

**Comment by FlorianRappl** [comment]
The API layer is used for the API stuff (hence the name, property `Globals`). It is read-only from the user space (scripting language), but is shadowed from the User layer ("global scope", property `Scope`), which can be modified from the user-space.

The hierarchy is as follows:

Globals/API :arrow_right: GlobalVariables/Scope :arrow_right: LocalVariables/Scope

This hierarchy is traversed in reverse order, i.e., if the code is currently run in a function we go from a local scope to the global scope to the API scope. If we still haven't found what we are looking for `null` is returned. The distinction between global scope (used for the user layer) and a local scope is that the global scope writes to variables in any case (independent if defined or not), while the local scope only writes to variables that are defined.

See (no test found, however, still written to global scope):

```C
test = 5; //equivalent to var test = 5; as we are in global scope
test // 5
```

Against (no test found in local scope, thus written to global scope):

```C
(() => { test = 5; })();
test // 5
```

In contrast to (test defined in local scope, thus written to local scope):

```C
(() => { var test = 5; })();
test // undefined
```

We could be more strict and require `var` to be used in global scope as well, however, this way things are a little bit easier for users (feels more direct and natural).

However, note that

```C
sin = cos;
sin == cos // true
```

We shadow the API layer `sin` with a user layer defined `sin`. In the future a call to `delete sin` would remove the user layer `sin`. The benefit of leaving the API layer untouched is that, e.g., `Globals` may be shared and that `Globals` may only be modified by C# code controlling the engine. The ultimate advantage, however, is that one can use a custom dictionary for the scope. In an IDE (e.g., see Sumerics) one can therefore immediately represent all variables without having to omit / check if one of the variables is in fact an API-defined function. So we get the distinction we look for: We can immediately separate between what the user added via scripting and what was inserted as API.

**Pull request closed without merging**